pkg/unilogger/logger.go

```diff
@@ -0,0 +1,268 @@
+	*logger
+
+	addSource *bool
+	level     *slog.Level
+	name      string
+
+	slogHandler *SlogHandler
+}
+
+type Options struct {
+	handlerOptions
+
+	AddSource bool
+	Level     slog.Level
+	Output    io.Writer
+
+	TimeFunc func(t time.Time) time.Time
+}
+
+func NewNop() *Logger {
+	return &Logger{
+		logger: slog.New(slog.NewTextHandler(io.Discard, &slog.HandlerOptions{})),
+	}
+}
+
+func NewLogger(opts Options) *Logger {
+	if opts.Output == nil {
+		opts.Output = os.Stdout
+	}
+
+	if opts.TimeFunc == nil {
+		opts.TimeFunc = func(t time.Time) time.Time {
+			return t
+		}
+	}
+
+	l := &Logger{
+		addSource: &opts.AddSource,
+		level:     &opts.Level,
+	}
+
+	handlerOpts := &slog.HandlerOptions{
+		AddSource: true,
+		Level:     l.level,
+		ReplaceAttr: func(_ []string, a slog.Attr) slog.Attr {
+			switch a.Key {
+			// skip standard fields
+			case slog.LevelKey, slog.MessageKey, slog.TimeKey:
+				return slog.Attr{}
+			case slog.SourceKey:
+				if !*l.addSource {
+					return slog.Attr{}
+				}
+
+				s := a.Value.Any().(*slog.Source)
+
+				dir, file := filepath.Split(s.File)
+
+				a.Value = slog.StringValue(fmt.Sprintf("%s:%d",
+					filepath.Join(filepath.Base(dir), file),
+					s.Line,
+				))
+			default:
+				key := strings.SplitN(a.Key, ";", 3)
+				if key[0] == "raw" {
+					a.Key = strings.Join(key[1:], ";")
+					raw := make(map[string]any, 1)
+
+					switch key[1] {
+					case "json":
+						if err := json.Unmarshal([]byte(a.Value.String()), &raw); err == nil {
+							a.Value = slog.AnyValue(raw)
+						}
+					case "yaml":
+						if err := yaml.Unmarshal([]byte(a.Value.String()), &raw); err == nil {
+							a.Value = slog.AnyValue(raw)
+						}
+					}
+				}
+			}
+
+			return a
+		},
+	}
+
+	l.slogHandler = NewHandler(opts.Output, handlerOpts, opts.TimeFunc)
+
+	l.logger = slog.New(l.slogHandler.WithAttrs(nil))
+
+	return l
+}
+
+func (l *Logger) SetLevel(level Level) {
+	*l.addSource = slog.Level(level) <= slog.LevelDebug
+
+	*l.level = slog.Level(level)
+}
+
+func (l *Logger) SetOutput(w io.Writer) {
+	l.slogHandler.w = w
+}
+
+func (l *Logger) Named(name string) *Logger {
+	currName := name
+	if l.name != "" {
+		currName = fmt.Sprintf("%s.%s", l.name, name)
+	}
+
+	return &Logger{
+		logger:    l.logger.With(slog.String("logger", currName)),
+		addSource: l.addSource,
+		level:     l.level,
+		name:      currName,
+	}
+}
+
+func (l *Logger) With(args ...any) *Logger {
+	return &Logger{
+		logger:    l.logger.With(args...),
+		addSource: l.addSource,
+		level:     l.level,
+		name:      l.name,
+	}
+}
+
+func (l *Logger) WithGroup(name string) *Logger {
+	return &Logger{
+		logger:    l.logger.WithGroup(name),
+		addSource: l.addSource,
+		level:     l.level,
+		name:      l.name,
+	}
+}
+
+func (l *Logger) Trace(msg string, args ...any) {
+	ctx := logContext.SetCustomKeyContext(context.Background())
+	ctx = logContext.SetStackTraceContext(ctx, getStack())
+
+	l.Log(ctx, LevelTrace.Level(), msg, args...)
+}
+
+// deprecated
```

Comment:
Please use suggestion, except "see" marks

Reply:
1) What *ln functions? Where are they from? 
2) There is no Print functions
3) There is no Panic functions?
4) We do not need interface for logger, because of ADR discussing.

Reply:
discussed